src/typechecker/typecheck.cpp

```diff
@@ -61,26 +61,43 @@ struct TypecheckHelper {
 	Frontend::CompileTimeEnvironment& env() { return tc.env(); }
 
 	TypeChecker& tc;
+
+	bool is_bound_to_env(MonoId var) {
+
+		for (auto& scope : env().scopes())
+			for (auto bound_type : scope.m_type_vars)
+				if (free_vars_of(bound_type).count(var) != 0)
+					return true;
```

Comment:
This whole thing has mixed levels of care put into it.

First `env().scopes()` reads a little strange to me, you are using encapsulation yet it feels just like regular code. 

Second, `m_type_vars` is not encapsulated.

Lastly, the predicate in the if statement is "low level" logic and could be extracted into a function, e.g. `appears_free_in(bound_type, var)`

Reply:
I'm not trying to hide the presence of `env`, only its location, because it will likely move into `TypecheckHelper` soon. Also, _it is_ regular code, wdym?

I don't think m_type_vars needs to be encapsulated: `Scope` is just a data holder. If anything should be encapsulated, it's scopes itself, but i haven't thought of a way to do that nicely yet.

Reply:
Ah, so encapsulation helps with location. That's a fair point. I can understand the rest as well.